packages/block-editor/src/store/selectors.js

```diff
@@ -1284,6 +1284,71 @@ export const hasInserterItems = createSelector(
 	],
 );
 
+/**
+ * Determines whether there is only one item that may be inserted.
+ *
+ * @param {Object}  state        Editor state.
+ * @param {?string} rootClientId Optional root client ID of block list.
+ *
+ * @return {boolean} True if there is one item available, false if zero or more than one.
+ */
+export const hasOneAllowedItem = ( state, rootClientId = null ) => {
+	if ( rootClientId ) {
+		const parentBlockListSettings = getBlockListSettings( state, rootClientId );
+		return get( parentBlockListSettings, [ 'allowedBlocks', 'length' ], 0 ) === 1;
+	}
+
+	return false;
+};
+
+/**
+ * Determines whether there is only one item that may be inserted.
+ *
+ * @param {Object}  state        Editor state.
+ * @param {?string} rootClientId Optional root client ID of block list.
+ *
+ * @return {string} The name of the allowed block.
+ */
+export const getOneAllowedItem = ( state, rootClientId = null ) => {
+	if ( ! rootClientId ) {
+		return false;
+	}
+	if ( ! hasOneAllowedItem( state, rootClientId ) ) {
+		return false;
+	}
+
+	const parentBlockListSettings = getBlockListSettings( state, rootClientId );
+	const name = get( parentBlockListSettings, [ 'allowedBlocks', 0 ], 0 );
+
+	return name;
+};
+
+/**
+ * Determines whether there is only one item that may be inserted.
+ *
+ * @param {Object}  state           		Editor state.
+ * @param {?string} clientId 				Block client ID.
+ * @param {?string} destinationRootClientId Root client ID of block list.
+ * @param {boolean} isAppender 	    		Determines if the block is added to a set of existing
+ * 											blocks in a list.
+ * @return {number} The insertion index.
+ */
+export function getInsertionIndex( state, clientId, destinationRootClientId, isAppender ) {
+	// If the clientId is defined, we insert at the position of the block.
+	if ( clientId ) {
+		return getBlockIndex( state, clientId, destinationRootClientId );
+	}
+
+	// If there a selected block, we insert after the selected block.
+	const end = getBlockSelectionEnd( state );
+	if ( ! isAppender && end ) {
```

Comment:
Logically speaking, we could avoid evaluating `end` if we know upfront that `isAppender` is true. But that's a minor observation, doesn't require action.